optimize/local.go

```diff
@@ -59,181 +56,176 @@ import (
-		// commanded, until convergence.
-
-		switch op {
-		case NoOperation:
-		case InitIteration:
-			panic("optimize: Method returned InitIteration")
+		result := <-results
+		switch result.Op {
 		case PostIteration:
-			panic("optimize: Method returned PostIteration")
-		case MajorIteration:
-			status = performMajorIteration(optLoc, loc, stats, startTime, settings)
-		case MethodDone:
-			statuser, ok := method.(Statuser)
-			if !ok {
-				panic("optimize: method returned MethodDone is not a Statuser")
-			}
-			status, err = statuser.Status()
-			if status == NotTerminated {
-				panic("optimize: method returned MethodDone but a NotTerminated status")
-			}
-		default: // Any of the Evaluation operations.
-			evaluate(p, loc, op, x)
-			updateEvaluationStats(stats, op)
-			status, err = checkEvaluationLimits(p, stats, settings)
-		}
-		if status != NotTerminated || err != nil {
-			return
-		}
-		if settings.Recorder != nil {
-			stats.Runtime = time.Since(startTime)
-			err = settings.Recorder.Record(loc, op, stats)
+			break Loop
+		default:
+			op, err := l.Method.Iterate(result.Location)
 			if err != nil {
-				if status == NotTerminated {
-					status = Failure
-				}
-				return status, err
+				l.status = Failure
+				l.err = err
+				l.exitFailure(operations, results, result)
+				return
 			}
-		}
-
-		op, err = method.Iterate(loc)
-		if err != nil {
-			status = Failure
-			return
+			result.Op = op
+			operations <- result
 		}
 	}
+	l.cleanup(operations, results)
 }
 
-func getDefaultMethod(p *Problem) Method {
-	if p.Grad != nil {
-		return &BFGS{}
+// exitFailure cleans up from a failure of the local method.
+func (l *localGlobal) exitFailure(operation chan<- GlobalTask, result <-chan GlobalTask, task GlobalTask) {
+	task.Op = MethodDone
+	operation <- task
+	task = <-result
+	if task.Op != PostIteration {
+		panic("task should have returned post iteration")
 	}
-	return &NelderMead{}
+	l.cleanup(operation, result)
 }
 
-// getStartingLocation allocates and initializes the starting location for the minimization.
-func getStartingLocation(p *Problem, method Method, initX []float64, stats *Stats, settings *Settings) (*Location, error) {
-	dim := len(initX)
-	loc := newLocation(dim, method)
-	copy(loc.X, initX)
+func (l *localGlobal) cleanup(operation chan<- GlobalTask, result <-chan GlobalTask) {
+	// Guarantee that result is closed before operation is closed.
+	for range result {
+	}
+	close(operation)
+}
 
-	if settings.UseInitialData {
-		loc.F = settings.InitialValue
-		if loc.Gradient != nil {
-			initG := settings.InitialGradient
-			if initG == nil {
+func (l *localGlobal) getStartingLocation(operation chan<- GlobalTask, result <-chan GlobalTask, task GlobalTask) Operation {
+	copy(task.X, l.InitX)
+	if l.Settings.UseInitialData {
+		task.F = l.Settings.InitialValue
+		if task.Gradient != nil {
+			g := l.Settings.InitialGradient
+			if task.Gradient != nil && g == nil {
```

Comment:
task.Gradient is already checked for nil above.

Reply:
Done.